diff --git a/OPS b/OPS
index c35f67043..13100f790 100644
--- a/OPS
+++ b/OPS
@@ -63,6 +63,7 @@ OP_DISPLAY_ADDRESS "display full address of sender"
 OP_DISPLAY_HEADERS "display message and toggle header weeding"
 OP_DISPLAY_MESSAGE "display a message"
 OP_EDIT_LABEL "add, change, or delete a message's label"
+OP_EDIT_EXPIRES "edit the message EXPIRES date"
 OP_EDIT_MESSAGE "edit the raw message"
 OP_EDITOR_BACKSPACE "delete the char in front of the cursor"
 OP_EDITOR_BACKWARD_CHAR "move the cursor one character to the left"
diff --git a/commands.c b/commands.c
index e7922107a..d80797e46 100644
--- a/commands.c
+++ b/commands.c
@@ -549,9 +549,9 @@ int mutt_select_sort (int reverse)
         * string.  Note that mutt now supports multiline prompts, so
         * it's okay for the translation to take up to three lines.
         */
-	_("Rev-Sort Date/Frm/Recv/Subj/tO/Thread/Unsort/siZe/sCore/sPam/Label?: ") :
-	_("Sort Date/Frm/Recv/Subj/tO/Thread/Unsort/siZe/sCore/sPam/Label?: "),
-	_("dfrsotuzcpl")))
+	_("Rev-Sort Date/Frm/Recv/Subj/tO/Thread/Unsort/siZe/sCore/sPam/Label/Expire?: ") :
+	_("Sort Date/Frm/Recv/Subj/tO/Thread/Unsort/siZe/sCore/sPam/Label/Expire?: "),
+	_("dfrsotuzcple")))
   {
   case -1: /* abort - don't resort */
     return -1;
@@ -599,6 +599,10 @@ int mutt_select_sort (int reverse)
   case 11: /* (l)abel */
     Sort = SORT_LABEL;
     break;
+
+  case 12: /* (e)xpires */
+    Sort = SORT_EXPIRES;
+    break;
   }
   if (reverse)
     Sort |= SORT_REVERSE;
@@ -1007,6 +1011,66 @@ int mutt_edit_content_type (HEADER *h, BODY *b, FILE *fp)
   return structure_changed;
 }
 
+static void _mutt_set_expires (HEADER *h, const char *exp, time_t expt)
+{
+	if (!h)
+		return;
+	if (!h->env->expires && !exp)
+		return;
+	if (h->env->expires && exp && !mutt_strcmp (h->env->expires, exp))
+		return;
+	if (h->env->expires)
+		FREE(&h->env->expires);
+	h->env->expires = safe_strdup(exp);
+	h->expires = expt;
+	h->changed = h->env->expires_changed = 1;
+}
+
+void mutt_edit_expires (HEADER *h)
+{
+	char buf[LONG_STRING] = "";
+	time_t exp;
+
+	if (h && h->env->expires)
+		strncpy(buf, h->env->expires, LONG_STRING);
+
+	if (mutt_get_field("Expires: ", buf, sizeof(buf), 0))
+		return;
+	
+	if (*buf)
+	{
+		char cmd[LONG_STRING];
+		FILE *out;
+		pid_t pid;
+
+		snprintf(cmd, sizeof(cmd), "date -Rd '%s'", buf);
+		if ((pid = mutt_create_filter(cmd, NULL, &out, NULL)) > 0)
+		{
+			size_t n = fread(buf, 1, sizeof(buf)-1, out);
+			safe_fclose(&out);
+			while (n && (buf[n-1] == '\n' || buf[n-1] == '\r'))
+				n--;
+			buf[n] = 0;
+
+			if (mutt_wait_filter(pid))
+			{
+				mutt_error(_("Invalid date format"));
+				return;
+			}
+		}
+	}
+
+	exp = *buf ? mutt_parse_date(buf, NULL) : 0;
+	if (h)
+		_mutt_set_expires(h, *buf ? buf : 0, exp);
+	else
+	{
+		int i;
+		for (i = 0; i < Context->vcount; i++)
+			if (Context->hdrs[Context->v2r[i]]->tagged)
+				_mutt_set_expires(Context->hdrs[Context->v2r[i]], *buf ? buf : 0, exp);
+	}
+}
 
 static int _mutt_check_traditional_pgp (HEADER *h, int *redraw)
 {
diff --git a/copy.c b/copy.c
index 7c49e7774..25fa8d7ea 100644
--- a/copy.c
+++ b/copy.c
@@ -106,6 +106,9 @@ mutt_copy_hdr (FILE *in, FILE *out, LOFF_T off_start, LOFF_T off_end, int flags,
 	    (ascii_strncasecmp ("Content-Length:", buf, 15) == 0 ||
 	     ascii_strncasecmp ("Lines:", buf, 6) == 0))
 	  continue;
+	if ((flags & CH_UPDATE_EXPIRES) &&
+	    ascii_strncasecmp ("Expires:", buf, 8) == 0)
+	  continue;
 	if ((flags & CH_UPDATE_REFS) &&
 	    ascii_strncasecmp ("References:", buf, 11) == 0)
 	  continue;
@@ -216,6 +219,9 @@ mutt_copy_hdr (FILE *in, FILE *out, LOFF_T off_start, LOFF_T off_end, int flags,
 	     ascii_strncasecmp ("type:", buf + 8, 5) == 0)) ||
 	   ascii_strncasecmp ("mime-version:", buf, 13) == 0))
 	continue;
+      if ((flags & CH_UPDATE_EXPIRES) &&
+	  ascii_strncasecmp ("Expires:", buf, 8) == 0)
+	continue;
       if ((flags & CH_UPDATE_REFS) &&
 	  ascii_strncasecmp ("References:", buf, 11) == 0)
 	continue;
@@ -356,8 +362,9 @@ mutt_copy_header (FILE *in, HEADER *h, FILE *out, int flags, const char *prefix)
 
   if (h->env)
     flags |= (h->env->irt_changed ? CH_UPDATE_IRT : 0)
-      | (h->env->refs_changed ? CH_UPDATE_REFS : 0);
-
+      | (h->env->refs_changed ? CH_UPDATE_REFS : 0)
+      | (h->env->expires_changed ? CH_UPDATE_EXPIRES : 0);
+  
   if (mutt_copy_hdr (in, out, h->offset, h->content->offset, flags, prefix) == -1)
     return -1;
 
@@ -457,6 +464,13 @@ mutt_copy_header (FILE *in, HEADER *h, FILE *out, int flags, const char *prefix)
         return -1;
   }
 
+  if (flags & CH_UPDATE_EXPIRES && h->env->expires)
+  {
+    fputs ("Expires: ", out);
+    fputs (h->env->expires, out);
+    fputc ('\n', out);
+  }
+
   if ((flags & CH_NONEWLINE) == 0)
   {
     if (flags & CH_PREFIX)
diff --git a/copy.h b/copy.h
index b47c7fbd0..5dd641c03 100644
--- a/copy.h
+++ b/copy.h
@@ -55,6 +55,7 @@
 #define CH_DISPLAY        (1<<18) /* display result to user */
 #define CH_UPDATE_LABEL   (1<<19) /* update X-Label: from hdr->env->x_label? */
 #define CH_VIRTUAL	  (1<<20) /* write virtual header lines too */
+#define CH_UPDATE_EXPIRES (1<<21) /* update Expires: */
 
 
 int mutt_copy_hdr (FILE *, FILE *, LOFF_T, LOFF_T, int, const char *);
diff --git a/curs_main.c b/curs_main.c
index 0776dd458..aaafa0cc1 100644
--- a/curs_main.c
+++ b/curs_main.c
@@ -2739,6 +2739,21 @@ int mutt_index_menu (void)
 
 	break;
 
+      case OP_EDIT_EXPIRES:
+
+	CHECK_MSGCOUNT;
+        CHECK_VISIBLE;
+	CHECK_READONLY;
+	CHECK_ACL(MUTT_ACL_WRITE, _("edit expiration"));
+	mutt_edit_expires(tag ? NULL : CURHDR);
+	if (menu->menu == MENU_PAGER)
+	{
+	  op = OP_DISPLAY_MESSAGE;
+	  continue;
+	}
+
+	break;
+
       case OP_FORWARD_MESSAGE:
 
 	CHECK_MSGCOUNT;
diff --git a/doc/Makefile.am b/doc/Makefile.am
index 34ce6247a..7924e4946 100644
--- a/doc/Makefile.am
+++ b/doc/Makefile.am
@@ -120,10 +120,10 @@ stamp-doc-rc: $(top_srcdir)/init.h makedoc$(EXEEXT) $(srcdir)/Muttrc.head
 	touch stamp-doc-rc
 
 manual.html: $(srcdir)/html.xsl $(srcdir)/mutt.xsl stamp-doc-xml $(srcdir)/mutt.css
-	-xsltproc --nonet -o $@ $(srcdir)/html.xsl manual.xml
+	-xsltproc -o $@ $(srcdir)/html.xsl manual.xml
 
 stamp-doc-chunked: $(srcdir)/chunk.xsl $(srcdir)/mutt.xsl stamp-doc-xml $(srcdir)/mutt.css
-	-xsltproc --nonet $(srcdir)/chunk.xsl manual.xml
+	-xsltproc $(srcdir)/chunk.xsl manual.xml
 	touch stamp-doc-chunked
 
 $(CHUNKED_DOCFILES): stamp-doc-chunked
diff --git a/functions.h b/functions.h
index e8a64bdd1..9ea8f1bbd 100644
--- a/functions.h
+++ b/functions.h
@@ -104,6 +104,7 @@ const struct binding_t OpMain[] = { /* map: index */
   { "edit",			OP_EDIT_MESSAGE,		"e" },
   { "edit-label",		OP_EDIT_LABEL,			"Y" },
   { "edit-type",		OP_EDIT_TYPE,			"\005" },
+  { "edit-expiration",		OP_EDIT_EXPIRES,		"E" },
   { "forward-message",		OP_FORWARD_MESSAGE,		"f" },
 #ifdef USE_NNTP
   { "forward-to-group",		OP_FORWARD_TO_GROUP,		"\033F" },
@@ -236,6 +237,7 @@ const struct binding_t OpPager[] = { /* map: pager */
   { "edit",		OP_EDIT_MESSAGE,		"e" },
   { "edit-label",	OP_EDIT_LABEL,			"Y" },
   { "edit-type",	OP_EDIT_TYPE,			"\005" },
+  { "edit-expiration",	OP_EDIT_EXPIRES,		"E" },
 #ifdef USE_NNTP
   { "followup-message",	OP_FOLLOWUP,			"F" },
   { "forward-to-group",	OP_FORWARD_TO_GROUP,		"\033F" },
diff --git a/hcache.c b/hcache.c
index cc5d6b20c..301b70659 100644
--- a/hcache.c
+++ b/hcache.c
@@ -526,6 +526,7 @@ dump_envelope(ENVELOPE * e, unsigned char *d, int *off, int convert)
   d = dump_char(e->message_id, d, off, 0);
   d = dump_char(e->supersedes, d, off, 0);
   d = dump_char(e->date, d, off, 0);
+  d = dump_char(e->expires, d, off, 0);
   d = dump_char(e->x_label, d, off, convert);
 
   d = dump_buffer(e->spam, d, off, convert);
@@ -569,6 +570,7 @@ restore_envelope(ENVELOPE * e, const unsigned char *d, int *off, int convert)
   restore_char(&e->message_id, d, off, 0);
   restore_char(&e->supersedes, d, off, 0);
   restore_char(&e->date, d, off, 0);
+  restore_char(&e->expires, d, off, 0);
   restore_char(&e->x_label, d, off, convert);
 
   restore_buffer(&e->spam, d, off, convert);
diff --git a/hdrline.c b/hdrline.c
index 7ec82c2b0..896dc288e 100644
--- a/hdrline.c
+++ b/hdrline.c
@@ -919,7 +919,9 @@ hdr_format_str (char *dest,
     
       ch = ' ';
 
-      if (WithCrypto && hdr->security & GOODSIGN)
+      if (hdr->expires && hdr->expires < time(NULL))
+	      ch = 'E';
+      else if (WithCrypto && hdr->security & GOODSIGN)
         ch = 'S';
       else if (WithCrypto && hdr->security & ENCRYPT)
       	ch = 'P';
diff --git a/init.h b/init.h
index 7dd7ebda8..e914dad3d 100644
--- a/init.h
+++ b/init.h
@@ -3771,6 +3771,11 @@ struct option_t MuttVars[] = {
   ** You may optionally use the ``reverse-'' prefix to specify reverse sorting
   ** order (example: ``\fCset sort_browser=reverse-date\fP'').
   */
+  { "sort_domain",	DT_BOOL, R_INDEX|R_RESORT, OPTSORTDOMAIN, 0 },
+  /*
+  ** .pp
+  ** When sorting by addresses (to or from), sort according to domain and user.
+  */
   { "sort_re",		DT_BOOL, R_INDEX|R_RESORT|R_RESORT_INIT, OPTSORTRE, 1 },
   /*
   ** .pp
@@ -4384,6 +4389,8 @@ const struct mapping_t SortMethods[] = {
   { "score",		SORT_SCORE },
   { "spam",		SORT_SPAM },
   { "label",		SORT_LABEL },
+  { "expiration",	SORT_EXPIRES },
+  { "status",		SORT_STATUS },
   { NULL,               0 }
 };
 
@@ -4404,6 +4411,8 @@ const struct mapping_t SortAuxMethods[] = {
   { "score",		SORT_SCORE },
   { "spam",		SORT_SPAM },
   { "label",		SORT_LABEL },
+  { "expiration",	SORT_EXPIRES },
+  { "status",		SORT_STATUS },
   { NULL,               0 }
 };
 
diff --git a/mbox.c b/mbox.c
index 346688323..4003c42e5 100644
--- a/mbox.c
+++ b/mbox.c
@@ -571,6 +571,7 @@ static int strict_cmp_envelopes (const ENVELOPE *e1, const ENVELOPE *e2)
   {
     if (mutt_strcmp (e1->message_id, e2->message_id) ||
 	mutt_strcmp (e1->subject, e2->subject) ||
+	mutt_strcmp (e1->expires, e2->expires) ||
 	!strict_cmp_lists (e1->references, e2->references) ||
 	!strict_addrcmp (e1->from, e2->from) ||
 	!strict_addrcmp (e1->sender, e2->sender) ||
diff --git a/mh.c b/mh.c
index bafed5406..619f227dd 100644
--- a/mh.c
+++ b/mh.c
@@ -1822,7 +1822,7 @@ static int mh_sync_message (CONTEXT * ctx, int msgno)
   HEADER *h = ctx->hdrs[msgno];
 
   if (h->attach_del || h->xlabel_changed ||
-      (h->env && (h->env->refs_changed || h->env->irt_changed)))
+      (h->env && (h->env->refs_changed || h->env->irt_changed || h->env->expires_changed)))
     if (mh_rewrite_message (ctx, msgno) != 0)
       return -1;
 
@@ -1834,7 +1834,7 @@ static int maildir_sync_message (CONTEXT * ctx, int msgno)
   HEADER *h = ctx->hdrs[msgno];
 
   if (h->attach_del || h->xlabel_changed ||
-      (h->env && (h->env->refs_changed || h->env->irt_changed)))
+      (h->env && (h->env->refs_changed || h->env->irt_changed || h->env->expires_changed)))
   {
     /* when doing attachment deletion/rethreading, fall back to the MH case. */
     if (mh_rewrite_message (ctx, msgno) != 0)
diff --git a/mutt.h b/mutt.h
index 6c9da6610..b72a11c77 100644
--- a/mutt.h
+++ b/mutt.h
@@ -508,6 +508,7 @@ enum
 #endif
   OPTSIGDASHES,
   OPTSIGONTOP,
+  OPTSORTDOMAIN,
   OPTSORTRE,
   OPTSPAMSEP,
   OPTSTATUSONTOP,
@@ -702,6 +703,7 @@ typedef struct envelope
   char *message_id;
   char *supersedes;
   char *date;
+  char *expires;
   char *x_label;
   char *organization;
 #ifdef USE_NNTP
@@ -715,6 +717,7 @@ typedef struct envelope
   LIST *in_reply_to;		/* in-reply-to header content */
   LIST *userhdrs;		/* user defined headers */
 
+  unsigned int expires_changed : 1;
   unsigned int irt_changed : 1; /* In-Reply-To changed to link/break threads */
   unsigned int refs_changed : 1; /* References changed to break thread */
 } ENVELOPE;
@@ -837,7 +840,6 @@ typedef struct header
   unsigned int attach_del : 1; 		/* has an attachment marked for deletion */
   unsigned int old : 1;
   unsigned int read : 1;
-  unsigned int expired : 1; 		/* already expired? */
   unsigned int superseded : 1; 		/* got superseded? */
   unsigned int replied : 1;
   unsigned int subject_changed : 1; 	/* used for threading */
@@ -874,6 +876,7 @@ typedef struct header
 
   time_t date_sent;     	/* time when the message was sent (UTC) */
   time_t received;      	/* time when the message was placed in the mailbox */
+  time_t expires;
   LOFF_T offset;          	/* where in the stream does this message begin? */
   int lines;			/* how many lines in the body of this message? */
   int index;			/* the absolute (unsorted) message number */
diff --git a/mutt_idna.c b/mutt_idna.c
index cb231b5a4..c4ac886ce 100644
--- a/mutt_idna.c
+++ b/mutt_idna.c
@@ -302,6 +302,41 @@ const char *mutt_addr_for_display (ADDRESS *a)
   return buff;
 }
 
+/* convert user@domain.com to com.domain!user for sorting purposes */
+const char *mutt_revaddr_for_sort (ADDRESS *a)
+{
+  static char *buff = NULL;
+  char *tmp = NULL;
+  char *user = NULL;
+  char *domain = NULL;
+  char *p, *q;
+  size_t dl;
+
+  FREE (&buff);
+
+  if (mbox_to_udomain (a->mailbox, &user, &domain) == -1)
+    return a->mailbox;
+  if ((tmp = intl_to_local (user, domain, MI_MAY_BE_IRREVERSIBLE)))
+    mbox_to_udomain (tmp, &user, &domain);
+
+  dl = mutt_strlen(domain);
+  safe_realloc (&buff, dl + mutt_strlen (user) + 2);
+
+  q = buff;
+  while ((p = memrchr(domain, '.', dl))) {
+    memcpy(q, p+1, domain+dl-p-1);
+    q += domain+dl-p-1;
+    dl = p-domain;
+    *q++ = '.';
+  }
+  memcpy(q, domain, dl);
+  q += dl;
+  FREE (&tmp);
+  *q++ = '!';
+  strcpy(q, user);
+  return buff;
+}
+
 /* Convert an ENVELOPE structure */
 
 void mutt_env_to_local (ENVELOPE *e)
diff --git a/mutt_idna.h b/mutt_idna.h
index 62e6e301d..197e96775 100644
--- a/mutt_idna.h
+++ b/mutt_idna.h
@@ -48,6 +48,8 @@
 # endif
 #endif /* defined(HAVE_LIBIDN) || defined(HAVE_LIBIDN2) */
 
+const char *mutt_revaddr_for_sort (ADDRESS *a);
+
 
 #ifdef HAVE_ICONV
 int mutt_addrlist_to_intl (ADDRESS *, char **);
diff --git a/muttlib.c b/muttlib.c
index 6036c2cf7..e3a51d8b2 100644
--- a/muttlib.c
+++ b/muttlib.c
@@ -752,6 +752,7 @@ void mutt_free_envelope (ENVELOPE **p)
   FREE (&(*p)->message_id);
   FREE (&(*p)->supersedes);
   FREE (&(*p)->date);
+  FREE (&(*p)->expires);
   FREE (&(*p)->x_label);
   FREE (&(*p)->organization);
 #ifdef USE_NNTP
@@ -788,6 +789,10 @@ void mutt_merge_envelopes(ENVELOPE* base, ENVELOPE** extra)
   MOVE_ELEM(message_id);
   MOVE_ELEM(supersedes);
   MOVE_ELEM(date);
+  if (!base->expires_changed)
+  {
+    MOVE_ELEM(expires);
+  }
   MOVE_ELEM(x_label);
   if (!base->refs_changed)
   {
diff --git a/parse.c b/parse.c
index 4b45ee138..ded60b9d5 100644
--- a/parse.c
+++ b/parse.c
@@ -1066,9 +1066,12 @@ int mutt_parse_rfc822_line (ENVELOPE *e, HEADER *hdr, char *line, char *p, short
     break;
     
     case 'e':
-    if (!ascii_strcasecmp ("xpires", line + 1) &&
-	hdr && mutt_parse_date (p, NULL) < time (NULL))
-      hdr->expired = 1;
+    if (!ascii_strcasecmp ("xpires", line + 1))
+    {
+      mutt_str_replace(&e->expires, p);
+      if (hdr)
+	hdr->expires = mutt_parse_date(p, NULL);
+    }
     break;
     
     case 'f':
diff --git a/pattern.c b/pattern.c
index 942ad83f2..029bdeb73 100644
--- a/pattern.c
+++ b/pattern.c
@@ -963,6 +963,10 @@ pattern_t *mutt_pattern_comp (/* const */ char *s, int flags, BUFFER *err)
 	    return NULL;
 	  }
 	}
+	else if (tmp->op == MUTT_EXPIRED)
+	{
+	  tmp->max = time(NULL);
+	}
 	implicit = 1;
 	break;
       case '(':
@@ -1219,7 +1223,7 @@ mutt_pattern_exec (struct pattern_t *pat, pattern_exec_flag flags, CONTEXT *ctx,
     case MUTT_ALL:
       return (!pat->not);
     case MUTT_EXPIRED:
-      return (pat->not ^ h->expired);
+      return (pat->not ^ (h->expires && h->expires < pat->max));
     case MUTT_SUPERSEDED:
       return (pat->not ^ h->superseded);
     case MUTT_FLAG:
diff --git a/protos.h b/protos.h
index 1bf95d6d0..bfa36efaa 100644
--- a/protos.h
+++ b/protos.h
@@ -190,6 +190,7 @@ void mutt_display_address (ENVELOPE *);
 void mutt_display_sanitize (char *);
 void mutt_draw_statusline (int cols, const char *buf, int buflen);
 int mutt_edit_content_type (HEADER *, BODY *, FILE *);
+void mutt_edit_expires (HEADER *);
 void mutt_edit_file (const char *, const char *);
 void mutt_edit_headers (const char *, const char *, HEADER *, char *, size_t);
 char **mutt_envlist (void);
diff --git a/sort.c b/sort.c
index 208dc3371..e11fa6432 100644
--- a/sort.c
+++ b/sort.c
@@ -38,6 +38,7 @@
 
 /* function to use as discriminator when normal sort method is equal */
 static sort_t *AuxSort = NULL;
+static time_t Now;
 
 #define AUXSORT(code,a,b) if (!code && AuxSort && !option(OPTAUXSORT)) { \
   set_option(OPTAUXSORT); \
@@ -112,6 +113,13 @@ const char *mutt_get_name (ADDRESS *a)
   return ("");
 }
 
+static const char *sort_addr (ADDRESS *a)
+{
+  if (option (OPTSORTDOMAIN) && a && a->mailbox)
+    return (mutt_revaddr_for_sort (a));
+  return (mutt_get_name (a));
+}
+
 static int compare_to (const void *a, const void *b)
 {
   HEADER **ppa = (HEADER **) a;
@@ -120,8 +128,8 @@ static int compare_to (const void *a, const void *b)
   const char *fb;
   int result;
 
-  strfcpy (fa, mutt_get_name ((*ppa)->env->to), SHORT_STRING);
-  fb = mutt_get_name ((*ppb)->env->to);
+  strfcpy (fa, sort_addr ((*ppa)->env->to), SHORT_STRING);
+  fb = sort_addr ((*ppb)->env->to);
   result = mutt_strncasecmp (fa, fb, SHORT_STRING);
   AUXSORT(result,a,b);
   return (SORTCODE (result));
@@ -135,8 +143,8 @@ static int compare_from (const void *a, const void *b)
   const char *fb;
   int result;
 
-  strfcpy (fa, mutt_get_name ((*ppa)->env->from), SHORT_STRING);
-  fb = mutt_get_name ((*ppb)->env->from);
+  strfcpy (fa, sort_addr ((*ppa)->env->from), SHORT_STRING);
+  fb = sort_addr ((*ppb)->env->from);
   result = mutt_strncasecmp (fa, fb, SHORT_STRING);
   AUXSORT(result,a,b);
   return (SORTCODE (result));
@@ -256,6 +264,34 @@ int compare_label (const void *a, const void *b)
   return (SORTCODE(result));
 }
 
+static int compare_expires (const void *a, const void *b)
+{
+  HEADER **pa = (HEADER **) a;
+  HEADER **pb = (HEADER **) b;
+  int result = 0;
+  if (!(*pa)->expires && !(*pb)->expires);
+  else if (!(*pa)->expires) result = 1;
+  else if (!(*pb)->expires) result = -1;
+  else result = (*pa)->expires - (*pb)->expires;
+  AUXSORT(result,a,b);
+  return (SORTCODE (result));
+}
+
+static int compare_status (const void *a, const void *b)
+{
+  HEADER **pa = (HEADER **) a;
+  HEADER **pb = (HEADER **) b;
+  int result = 0;
+  if (!result) result = (*pb)->deleted - (*pa)->deleted;
+  if (!result) result = (*pb)->old - (*pa)->old;
+  if (!result) result = (*pb)->read - (*pa)->read;
+  if (!result) result = (*pa)->flagged - (*pb)->flagged;
+  if (!result) result = ((*pb)->expires && (*pb)->expires < Now) - ((*pa)->expires && (*pa)->expires < Now);
+  if (!result) result = (*pb)->replied - (*pa)->replied;
+  AUXSORT(result,a,b);
+  return (SORTCODE(result));
+}
+
 sort_t *mutt_get_sort_func (int method)
 {
   switch (method & SORT_MASK)
@@ -280,6 +316,10 @@ sort_t *mutt_get_sort_func (int method)
       return (compare_spam);
     case SORT_LABEL:
       return (compare_label);
+    case SORT_EXPIRES:
+      return (compare_expires);
+    case SORT_STATUS:
+      return (compare_status);
     default:
       return (NULL);
   }
@@ -328,6 +368,8 @@ void mutt_sort_headers (CONTEXT *ctx, int init)
   if (init && ctx->tree)
     mutt_clear_threads (ctx);
 
+  Now = time(NULL);
+
   if ((Sort & SORT_MASK) == SORT_THREADS)
   {
     AuxSort = NULL;
diff --git a/sort.h b/sort.h
index b2c35cfd4..84a66c420 100644
--- a/sort.h
+++ b/sort.h
@@ -37,6 +37,8 @@
 #define SORT_PATH	18
 #define SORT_LABEL	19
 #define SORT_DESC	20
+#define SORT_EXPIRES	21
+#define SORT_STATUS	22
 
 /* Sort and sort_aux are shorts, and are a composite of a
  * constant sort operation number and a set of compounded
