From a760a2341322d0c5c8cb79cbc5ee198e749d98ad Mon Sep 17 00:00:00 2001
From: Dylan Simon <dylan@dylex.net>
Date: Thu, 22 Jul 2010 16:27:12 -0400
Subject: [PATCH 2/2] support command-line specified password files

---
 xtrlock.c |   34 ++++++++++++++++++++++++++++++----
 1 files changed, 30 insertions(+), 4 deletions(-)

diff --git a/xtrlock.c b/xtrlock.c
index 137e212..bd94ee4 100644
--- a/xtrlock.c
+++ b/xtrlock.c
@@ -53,7 +53,8 @@ Window window, root;
 #define INITIALGOODWILL MAXGOODWILL
 #define GOODWILLPORTION 0.3
 
-struct passwd *pw;
+static char spw[256];
+static char fpw[256];
 int passwordok(const char *s) {
 #if 0
   char key[3];
@@ -67,7 +68,11 @@ int passwordok(const char *s) {
 #else
   /* simpler, and should work with crypt() algorithms using longer
      salt strings (like the md5-based one on freebsd).  --marekm */
-  return !strcmp(crypt(s, pw->pw_passwd), pw->pw_passwd);
+  if (*spw && !strcmp(crypt(s, spw), spw))
+	  return 1;
+  if (*fpw && !strcmp(crypt(s, fpw), fpw))
+	  return 1;
+  return 0;
 #endif
 }
 
@@ -82,11 +87,12 @@ int main(int argc, char **argv){
   Pixmap csr_source,csr_mask;
   XColor csr_fg, csr_bg, dummy;
   int ret;
+  struct passwd *pw;
 #ifdef SHADOW_PWD
   struct spwd *sp;
 #endif
 
-  if (argc != 1) {
+  if (argc > 2) {
     fprintf(stderr,"xtrlock (version %s): no arguments allowed\n",program_version);
     exit(1);
   }
@@ -108,7 +114,27 @@ int main(int argc, char **argv){
      and we don't need root privileges any longer.  --marekm */
   setuid(getuid());
 
-  if (strlen(pw->pw_passwd) < 13) {
+  if (argc == 2)
+  {
+	  FILE *f = strcmp(argv[1], "-") ? fopen(argv[1], "r") : stdin;
+	  if (!f)
+		  perror("specified password file");
+	  else
+	  {
+		  char *p;
+		  fgets(fpw, 256, f);
+		  fclose(f);
+		  if (p = strchr(fpw, '\n'))
+			  *p = 0;
+		  if (strlen(fpw) < 13)
+			  *fpw = 0;
+	  }
+  }
+
+  if (strlen(pw->pw_passwd) >= 13) {
+	  strncpy(spw, pw->pw_passwd, 255);
+  }
+  else if (!*fpw) {
     fputs("password entry has no pwd\n",stderr); exit(1);
   }
   
-- 
1.7.1

